File: clients/python/examples/with_panic_handler.py
"""Demonstrates ``varta.panic.install_excepthook_uds``.

Any uncaught exception emits a single ``Status.CRITICAL`` frame with
``nonce=NONCE_TERMINAL`` so the observer can distinguish a structured
panic from a regular Critical beat.

Run alongside varta-watch::

    varta-watch --socket /tmp/varta.sock --threshold-ms 2000 &
    python with_panic_handler.py /tmp/varta.sock
"""

from __future__ import annotations

import sys
import time

from varta import Status, Varta
from varta.panic import install_excepthook_uds


def main(argv: list[str]) -> int:
    path = argv[1] if len(argv) > 1 else "/tmp/varta.sock"
    install_excepthook_uds(path)

    with Varta.connect(path) as agent:
        agent.beat(Status.OK)
        time.sleep(0.1)
        # Force an uncaught exception. The installed excepthook emits a
        # critical beat before the interpreter prints the traceback.
        raise RuntimeError("simulated agent panic")
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: crates/varta-client/examples/with_panic_handler.rs
//! Demonstrates the opt-in panic handler.
//!
//! When the process panics, a `Status::Critical` VLP frame is emitted to the
//! observer before the normal panic message and stack unwind proceed.
//!
//! Build and run:
//!
//! ```sh
//! varta-watch --socket /tmp/varta.sock --threshold-ms 2000 &
//! cargo run --example with_panic_handler --features panic-handler
//! ```

fn main() -> std::io::Result<()> {
    // Register the hook before connecting so any early panic is still signalled.
    varta_client::install_panic_handler("/tmp/varta.sock");

    let mut agent = varta_client::Varta::connect("/tmp/varta.sock")?;
    for _ in 0..10 {
        let _ = agent.beat(varta_client::Status::Ok, 0);
        std::thread::sleep(std::time::Duration::from_millis(200));
    }
    Ok(())
}
